hadoop-tools/hadoop-aws/src/test/java/org/apache/hadoop/fs/s3a/ITestS3AIOStatisticsContext.java

```diff
@@ -0,0 +1,206 @@
+            }
+            latch.countDown();
+          } catch (Exception e) {
+            latch.countDown();
+            setFutureException(e);
+            LOG.error("An error occurred while doing a task in the thread", e);
+          } catch (AssertionError ase) {
+            latch.countDown();
+            setFutureASE(ase);
+            throw ase;
+          }
+        });
+      }
+      // wait for tasks to finish.
+      latch.await();
+    } finally {
+      executor.shutdown();
+    }
+
+    // Check if an Excp or ASE was caught while the test threads were running.
+    maybeReThrowFutureException();
+    maybeReThrowFutureASE();
+
+  }
+
+  @Test
+  public void testS3ABlockOutputStreamIOStatisticsContext()
+      throws Exception {
+    S3AFileSystem fs = getFileSystem();
+    Path path = path(getMethodName());
+    byte[] data = dataset(256, 'a', 'z');
+    byte[] writeDataFirst = new byte[READ_BYTES_FIRST];
+    byte[] writeDataSecond = new byte[READ_BYTES_SECOND];
+    writeDataset(fs, path, data, data.length, 1024, true);
+
+    final ExecutorService executor =
+        HadoopExecutors.newFixedThreadPool(SMALL_THREADS);
+    CountDownLatch latch = new CountDownLatch(SMALL_THREADS);
+
+    try {
+      for (int i = 0; i < SMALL_THREADS; i++) {
+        executor.submit(() -> {
+          try {
+            IOStatistics ioStatisticsFirst;
+            try (FSDataOutputStream out = fs.create(path)) {
+              out.write(writeDataFirst);
+              out.close();
+              ioStatisticsFirst = assertThreadStatisticsBytesWrite(out,
+                  READ_BYTES_FIRST);
+            }
+            // Stream is closed for a thread. Re-open and do more operations.
+            IOStatistics ioStatisticsSecond;
+            try (FSDataOutputStream out = fs.create(path)) {
+              out.write(writeDataSecond);
+              out.close();
+              ioStatisticsSecond = assertThreadStatisticsBytesWrite(out,
+                  READ_BYTES_FIRST + READ_BYTES_SECOND);
+            }
+            latch.countDown();
+          } catch (Exception e) {
+            latch.countDown();
+            setFutureException(e);
+            LOG.error("An error occurred while doing a task in the thread", e);
+          } catch (AssertionError ase) {
+            latch.countDown();
+            setFutureASE(ase);
+            throw ase;
+          }
+        });
+      }
+      // wait for tasks to finish.
+      latch.await();
+    } finally {
+      executor.shutdown();
+    }
+
+    // Check if an Excp or ASE was caught while the test threads were running.
+    maybeReThrowFutureException();
+    maybeReThrowFutureASE();
+
+  }
+
+  private IOStatistics assertThreadStatisticsBytesWrite(FSDataOutputStream out, int writeBytes) {
+    S3ABlockOutputStream s3aOut = (S3ABlockOutputStream) out.getWrappedStream();
+    IOStatisticsContext ioStatisticsContext = s3aOut.getIoStatisticsContext();
```

Comment:
stream should just cache the thread IOStatistics in its constructor, not the context.